# Analysis_APM

APM model is aimed to partition the genetic variance components of human complex traits due to additive, parent-of origin and maternal genetic effects based on mother-child genotypes.

The R scripts are used to:
+ run a set of simulations of APM model with or without the fixed effects 
+ compare REML, PCGC and R-PCGC in the estimation with case-control studies
+ generate the GRMs in the real data analysis
+ hold the functions to be used in simulating data which are cited from [Golan et al. (2014)](#Golan2014), [Laurin et al. (2018)](#Laurin2018), and [Yang et al. (2011)](#Yang2011)


# References

<div id="Golan2014"></div>
[1] Golan, D., Lander, E. S., and Rosset, S. (2014). Measuring missing heritability: inferring the contribution of common variants. Proceedings of the National Academy of Sciences of the United States of America, 111(49):E5272–81 (www.pnas.org/lookup/suppl/doi:10.1073/pnas.1419064111/-/DCSupplemental)

<div id="Laurin2018"></div>
[2] Laurin, C., Cuellar-Partida, G., Hemani, G., Smith, G., Yang, J.,  and Evans, D. (2018). Partitioning phenotypic variance due  to parent-of-origin effects using genomic relatednessmatrices. Behavior Genetics, 48:67–79 (https://link.springer.com/article/10.1007/s10519-017-9880-0)

<div id="Yang2011"></div>
[3] Yang, J., Lee, S. H., Goddard, M. E., and Visscher, P. M. (2011). GCTA: A tool for genome-wide complex trait analysis. The American Journal of Human Genetics, 88(1):76–82 (https://dx.doi.org/10.1016/j.ajhg.2010.11.011)
